### PR TITLE
Add Westmint Runtime

### PR DESCRIPTION
**Pull request by joepetrowski**
Follow on to https://github.com/paritytech/statemint/pull/62. Runtime with sudo in it to use on Westend.

Closes https://github.com/paritytech/statemint/issues/63

Weight files copied from Statemint.

### CODE REVIEW

File: runtime/westmint/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,823 @@
...
+/// to even the core data structures.
+pub mod opaque {
+	use super::*;
+
+	pub use sp_runtime::OpaqueExtrinsic as UncheckedExtrinsic;
+
+	/// Opaque block type.
+	pub type Block = generic::Block<Header, UncheckedExtrinsic>;
+
+	pub type SessionHandlers = ();
```
Comment (kianenigma) [inline runtime/westmint/src/lib.rs:92]: not sure what this is. Maybe I've added it myself. Either way, should probably go away.

File: runtime/westmint/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,823 @@
...
+
+	impl_opaque_keys! {
+		pub struct SessionKeys {
+			pub aura: Aura,
+		}
+	}
```
Comment (kianenigma) [inline runtime/westmint/src/lib.rs:98]:
```suggestion
	use super::*;

	pub use sp_runtime::OpaqueExtrinsic as UncheckedExtrinsic;

	/// Opaque block header type.
	pub type Header = generic::Header<BlockNumber, BlakeTwo256>;
	/// Opaque block type.
	pub type Block = generic::Block<Header, UncheckedExtrinsic>;
	/// Opaque block identifier type.
	pub type BlockId = generic::BlockId<Block>;

	impl_opaque_keys! {
		pub struct SessionKeys {
			pub aura: Aura,
		}
	}
```

Comment (kianenigma) [reply runtime/westmint/src/lib.rs:98]: would be nice to update the other two runtimes as well

File: runtime/westmint/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,823 @@
...
+	/// What to do if an account is fully reaped from the system.
+	type OnKilledAccount = ();
+	/// The data to be stored in an account.
+	type AccountData = pallet_balances::AccountData<Balance>;
+	/// Weight information for the extrinsics of this pallet.
+	type SystemWeightInfo = ();
+	/// This is used as an identifier of the chain. 42 is the generic substrate prefix.
+	type SS58Prefix = SS58Prefix;
+	/// The set code logic of the parachain.
+	type OnSetCode = cumulus_pallet_parachain_system::ParachainSetCode<Self>;
```
Comment (kianenigma) [inline runtime/westmint/src/lib.rs:195]:
```suggestion
	type BaseCallFilter = ();
	type BlockWeights = RuntimeBlockWeights;
	type BlockLength = RuntimeBlockLength;
	type AccountId = AccountId;
	type Call = Call;
	type Lookup = AccountIdLookup<AccountId, ()>;
	type Index = Index;
	type BlockNumber = BlockNumber;
	type Hash = Hash;
	type Hashing = BlakeTwo256;
	type Header = generic::Header<BlockNumber, BlakeTwo256>;
	type Event = Event;
	type Origin = Origin;
	type BlockHashCount = BlockHashCount;
	type DbWeight = RocksDbWeight;
	type Version = Version
	type PalletInfo = PalletInfo;
	type OnNewAccount = ();
	type OnKilledAccount = ();
	type AccountData = pallet_balances::AccountData<Balance>;
	type SystemWeightInfo = ();
	type SS58Prefix = SS58Prefix;
	type OnSetCode = cumulus_pallet_parachain_system::ParachainSetCode<Self>;
```

Similarly, I wanted to remove these in my cleanup PR as well. This is not an educational chain to have all of these. If you agree, reflecting in other two would be nice.

File: runtime/westmint/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,823 @@
...
+	type Trader = FixedRateOfConcreteFungible<WeightPrice, ()>;
+	type ResponseHandler = ();	// Don't handle responses for now.
+}
+
+parameter_types! {
+	pub const MaxDownwardMessageWeight: Weight = MAXIMUM_BLOCK_WEIGHT / 10;
+}
+
+/// No local origins on this chain are allowed to dispatch XCM sends/executions.
+pub type LocalOriginToLocation = ();
```
Comment (kianenigma) [inline runtime/westmint/src/lib.rs:526]: You need to make sure that you have reflected the changes of #60

**Review by kianenigma — COMMENTED** [review]
LGTM, but I think it is better to first merge #60 and then copy-paste the runtime over and modify, to make sure no aura dependency is being missed.

**Review by JesseAbram — COMMENTED** [review]
Code wise looks fine but 2 things. Will this require the weights be maintained and updated (if so there is a bash script called benchmarks.sh to update so we can run them) preferably if not can we just use the weights from statemint/mine. 

Also second, adding more runtimes means more things to update and maintain, if we are on westend, would we be able to use sudo to send updates down to statemint without any changes to the runtime? This would just be less code in general @apopiak has been working on the gov stuff and may have some insight?

**Comment by joepetrowski** [comment]
TODO: Update based on #50 to take upgrades from Westend sudo.

**Comment by apopiak** [comment]
I think we should probably extract the common parts into commons so that there is less redundant code to update

**Comment by apopiak** [comment]
> Will this require the weights be maintained and updated (if so there is a bash script called benchmarks.sh to update so we can run them) preferably if not can we just use the weights from statemint/mine.

It's a test runtime, so it's fine to just use statemint weights (or probably even eye-balled values).
 
> Also second, adding more runtimes means more things to update and maintain, if we are on westend, would we be able to use sudo to send updates down to statemint without any changes to the runtime? This would just be less code in general @apopiak has been working on the gov stuff and may have some insight?

I think it makes sense to have separate runtimes for separate relay chains.
As noted above I would prefer using techniques like moving stuff to the common crate (or other stuff like cargo features) to remove redundancy.

**Review by apopiak — APPROVED** [review]

File: runtime/westmint/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,812 @@
...
+	type ReservedXcmpWeight = ReservedXcmpWeight;
+}
+
+impl parachain_info::Config for Runtime {}
+
+impl cumulus_pallet_aura_ext::Config for Runtime {}
+
+parameter_types! {
+	pub const WestendLocation: MultiLocation = MultiLocation::X1(Junction::Parent);
+	pub const RelayNetwork: NetworkId = NetworkId::Polkadot;
```
Comment (joepetrowski) [inline runtime/westmint/src/lib.rs:415]:
```suggestion
	pub const RelayNetwork: NetworkId = NetworkId::Westend;
```

File: node/src/command.rs [hunk 6]
```diff
@@ -46,6 +46,8 @@ fn load_spec(
 		)?),
 		"statemine-dev" => Box::new(chain_spec::statemine_development_config(para_id)),
 		"statemine-local" => Box::new(chain_spec::statemine_local_config(para_id)),
+		"westmint-dev" => Box::new(chain_spec::westmint_development_config(para_id)),
+		"westmint-local" => Box::new(chain_spec::westmint_local_config(para_id)),
```
Comment (apopiak) [inline node/src/command.rs:50]: what about `statemine` and `westmint` specs?

Comment (joepetrowski) [reply node/src/command.rs:50]: Tracking here: https://github.com/paritytech/statemint/issues/14

**Pull request merged**